### PR TITLE
Crudspec

### PR DESCRIPTION
**Pull request by wtrocki**
Very very early version of crudspec

@craicoverflow can you accept invite to org.

### CODE REVIEW

File: docs/spec-subscriptions.md [hunk 1]
```diff
@@ -1,7 +1,21 @@
...
 ---
 
-spec-subscriptions TODO
+Subscriptions are divided to 3 different groups of changes
+Create, Update and Delete subscriptions that users can use with filtering capabilities
+
+``` graphql
+input NoteSubscriptionFilter {
+  id: ID
+  title: String
```
Comment (machi1990) [inline docs/spec-subscriptions.md:12]: Why is the type `String` and not `StringInput` ?

File: docs/spec-delete.md [hunk 2]
```diff
@@ -1,7 +1,56 @@
...
+
+```graphql
+type Mutation {
+  deleteNote(input: DeleteNoteInput!, where: DeleteNoteFilter): Note!
+}
+```
+Input type for filtering has all fields marked as optional
+```graphql
+input DeleteNoteFilter {
+  title: String
```
Comment (machi1990) [inline docs/spec-delete.md:53]: should the type be `StringInput`?

File: docs/concepts.md [hunk 3]
```diff
@@ -16,11 +16,14 @@ various operations that can be executed on the set of the objects:
...
-
+- Delete: delete entire object by id 
+- Get: get specific object by id
+- Find: find multiple objects using query
 
 ## Input types
 
-GraphQL CRUD defines common input type categories that can be used in various CRUD methods to define operations
+GraphQL CRUD defines common input type categories that can be used in various CRUD methods to define operations.
+For example Create operation will use specific input type that do not require object ID.
```
Comment (craicoverflow) [inline docs/concepts.md:26]:
```suggestion
For example Create operation will use specific input type that does not require object ID.
```

File: docs/concepts.md [hunk 4]
```diff
@@ -42,4 +45,4 @@ of the queries and mutations that can be used for different needs etc.
 GraphQL CRUD provides various extensions for more advanced use cases like:
 
 - Querying Server Side Delta
-- Subscriptions filtering
\ No newline at end of file
+- Subscriptions filtering etc.
```
Comment (craicoverflow) [inline docs/concepts.md:48]: "etc" is vague. I'd remove it unless you are going to explain it.

File: docs/getting-started.md [hunk 5]
```diff
@@ -8,21 +8,47 @@ sidebar_label: Getting Started
 
 GraphQL CRUD is an superset of the GraphQL giving developers
```
Comment (craicoverflow) [inline docs/getting-started.md:9]:
```suggestion
GraphQL CRUD is a superset of the GraphQL giving developers
```

File: docs/spec-getOne.md [hunk 6]
```diff
@@ -1,7 +1,14 @@
...
-title: spec-getOne
-sidebar_label: spec-getOne
+title: Get Operation
+sidebar_label: Get Operation
 ---
 
-spec-getOne TODO
+## Get single object 
+
+Fetching the note by id can be enabled by specifying id as element.
```
Comment (machi1990) [inline docs/spec-getOne.md:9]:
```suggestion
Fetching an object by id can be enabled by specifying id as element. 
```

File: docs/spec-getOne.md [hunk 7]
```diff
@@ -1,7 +1,14 @@
...
+title: Get Operation
+sidebar_label: Get Operation
 ---
 
-spec-getOne TODO
+## Get single object 
+
+Fetching the note by id can be enabled by specifying id as element. 
+Id can be mapped to id of the object on the resolver level.
+
```
Comment (machi1990) [inline docs/spec-getOne.md:11]:
```suggestion
For the Note type, this can be:
```

File: docs/getting-started.md [hunk 8]
```diff
@@ -8,21 +8,47 @@ sidebar_label: Getting Started
...
 patterns for building standard data driven GraphLQL APIs.
```
Comment (craicoverflow) [inline docs/getting-started.md:10]:
```suggestion
patterns for building standard data driven GraphQL APIs.
```

File: docs/getting-started.md [hunk 9]
```diff
@@ -8,21 +8,47 @@ sidebar_label: Getting Started
...
+  "user": {
+    "name": "Mark Zuckerberg"
+  }
+}
+```
+
+## Rationale
+
+GraphQL is a flexible query language supporting many different data access patterns. 
+For most of the projects CRUD operations turn out to be a very common pattern. Definifing this very common pattern enables the community to build tooling specific to the common CRUD style API.
```
Comment (craicoverflow) [inline docs/getting-started.md:35]:
```suggestion
For most of the projects CRUD operations turn out to be a very common pattern. Defining this very common pattern enables the community to build tooling specific to the common CRUD style API.
```

File: docs/spec-update.md [hunk 10]
```diff
@@ -1,7 +1,55 @@
...
+## Variations
+
+### Contidional updates
+
+Conditional updates can be enabled for the cases where we want to perform update 
+operation only after meeting certain criteria
+
+```graphql
+type Mutation {
+  updateNote(input: MutateNoteInput!, where: UpdateNoteFilter): Note!
```
Comment (machi1990) [inline docs/spec-update.md:46]: Where does the `where: UpdateNoteFilter ` coming from? 

Is this the same as the `filter` here https://github.com/graphqlcrud/spec/pull/6/files#diff-6c9e4cac1fb75081a4efb6eaa188da54R67?

File: docs/getting-started.md [hunk 11]
```diff
@@ -8,21 +8,47 @@ sidebar_label: Getting Started
...
 
 ## Targets of this spec
```
Comment (craicoverflow) [inline docs/getting-started.md:37]: Mixing terminology - is it a spec or is it not?

File: docs/getting-started.md [hunk 12]
```diff
@@ -8,21 +8,47 @@ sidebar_label: Getting Started
...
-3) Avoid corner cases or specifics of individual CRUD implementations 
-4) Define lenient standard based on practices, existing API and providers
+2. Provide overview of the data access methods that are used to fetch data
+3. Avoid corner cases or specifics of individual CRUD implementations 
+4. Define lenient standard based on practices, existing API and providers
 that do not enforce specific query or mutation names etc.
-5) Provide set of the reference implementations for *different programming languages* in case you want to adopt reference
+5. Provide set of the reference implementations for *different programming languages* in case you want to adopt reference
 implementation
+6. Provide capabilities available in vanila GraphQL (no preprocessors, helpers, annotations etc.)
```
Comment (craicoverflow) [inline docs/getting-started.md:47]:
```suggestion
6. Provide capabilities available in vanilla GraphQL (no preprocessors, helpers, annotations etc.)
```

File: docs/getting-started.md [hunk 13]
```diff
@@ -8,21 +8,47 @@ sidebar_label: Getting Started
...
 
 ## Non targets of this spec
 
-1) Define every possible method to map database or SQL Capabilities
+1. Define every possible method to map database or SQL Capabilities
 
 Over the years we seen issues with developers adopting very open CRUD capabilities on the client. 
 There is no silverbullet that will give developers both flexiblity of the query capabilities on the client
```
Comment (craicoverflow) [inline docs/getting-started.md:54]:
```suggestion
There is no silver bullet that will give developers both flexiblity of the query capabilities on the client 
```

File: docs/getting-started.md [hunk 14]
```diff
@@ -41,30 +67,4 @@ GraphQL CRUD focuses only on providing CRUD capaiblities.
 For general rules for GraphQL schemas please ses: https://graphql-rules.com
```
Comment (craicoverflow) [inline docs/getting-started.md:67]:
```suggestion
For general rules for GraphQL schemas see [GraphQL Rules](https://graphql-rules.com)
```

File: docs/reference-impl.md [hunk 15]
```diff
@@ -10,20 +10,13 @@ GraphQL CRUD can be applied out of the box using https://graphback.dev
 
 ### Node.js 
 
-Graphback provides ability to generate schema that will be fully compatible with the 
+Graphback (https://graphback.dev) provides ability to generate schema that will be fully compatible with the
```
Comment (craicoverflow) [inline docs/reference-impl.md:13]:
```suggestion
[Graphback](https://graphback.dev) provides ability to generate schema that will be fully compatible with the 
```

File: docs/reference-impl.md [hunk 16]
```diff
@@ -10,20 +10,13 @@ GraphQL CRUD can be applied out of the box using https://graphback.dev
...
 ### Node.js 
 
-Graphback provides ability to generate schema that will be fully compatible with the 
+Graphback (https://graphback.dev) provides ability to generate schema that will be fully compatible with the 
 GraphQLCRUD and also connect it directly to Postgress, MongoDB and other datasources without writing any code.
 
 ### JavaScript
 
-https://offix.dev provides reference implementation for the delta sync queries and implements GraphQL CRUD on the client side.
+Offix (https://offix.dev) provides reference implementation for the delta sync queries and implements GraphQL CRUD on the client side.
```
Comment (craicoverflow) [inline docs/reference-impl.md:18]:
```suggestion
[Offix](https://offix.dev) provides reference implementation for the delta sync queries and implements GraphQL CRUD on the client side.
```

File: docs/reference-impl.md [hunk 17]
```diff
@@ -33,10 +26,12 @@ Specification was build based on numerous community implementations
 - Prisma 
 - Hasura 
 - Postgraphile
+- SQLMancer
+- TypeORM
 - GraphCMS
-- GraphQL-CLI
-- GraphQL-Serve
-- GraphQL-Create
+- GraphQL-cli
```
Comment (craicoverflow) [inline docs/reference-impl.md:32]:
```suggestion
- GraphQL CLI
```

File: docs/schema.graphql [hunk 18]
```diff
@@ -0,0 +1,158 @@
+## NOTE: This schema was generated by Graphback and should not be changed manually
```
Comment (craicoverflow) [inline docs/schema.graphql:1]: Should remove Graphback references in main doc to remain impartial.

File: docs/spec-delete.md [hunk 19]
```diff
@@ -1,7 +1,56 @@
 ---
 id: spec-delete
-title: spec-delete
-sidebar_label: spec-delete
+title: Delete operation
 ---
 
-spec-delete TODO
+## Delete Specification
```
Comment (craicoverflow) [inline docs/spec-delete.md:6]: Delete Operation?

File: docs/spec-delete.md [hunk 20]
```diff
@@ -1,7 +1,56 @@
...
+```graphql
+type Note {
+  id: ID!
+  title: String!
+  description: String
+  comments: [Comment]!
+}
+```
+
+Following mutation will be created:
```
Comment (craicoverflow) [inline docs/spec-delete.md:20]: Implies generation, but doesn't have to be the case if someone wants to implement this in their own API with framework.

File: docs/spec-delete.md [hunk 21]
```diff
@@ -1,7 +1,56 @@
...
+## Variations
+
+### Contidional deletion
+
+Conditional deletion can be enabled for the cases where we want to perform deletion 
+operation only after meeting certain criteria
+
+```graphql
+type Mutation {
+  deleteNote(input: DeleteNoteInput!, where: DeleteNoteFilter): Note!
```
Comment (craicoverflow) [inline docs/spec-delete.md:47]: Previous example uses MutateNoteInput and this one uses DeleteNoteInput.

Intentional?

File: docs/spec-delete.md (hunk 2 above)
Comment (craicoverflow) [inline docs/spec-delete.md:53]: Should this have `title: StringFilterInput` etc?

File: docs/spec-find.md [hunk 22]
```diff
@@ -1,7 +1,84 @@
...
 ---
 
-spec-find TODO
+## Find multiple objects Specification
+
+Find operation can be respresented as follows
+
+```graphql
+type Query {
+  findNotes(filter: NoteFilter, orderBy: OrderByInput): [Note]!
```
Comment (craicoverflow) [inline docs/spec-find.md:12]: Should return NoteResultList

File: docs/spec-find.md [hunk 23]
```diff
@@ -1,7 +1,84 @@
...
+}
+```
+
+Finding objects can be done by creating numerous helpers.
+Filter input type:
+```graphql
+input NoteFilter {
```
Comment (craicoverflow) [inline docs/spec-find.md:19]: I think the example model uses should have a field with each of the common scalar types, ID, Int, Float, Boolean, String.

File: docs/spec-find.md [hunk 24]
```diff
@@ -1,7 +1,84 @@
...
+  id: IDInput
+  title: StringInput
+  description: StringInput
+  and: [NoteFilter]
+  or: [NoteFilter]
+  not: NoteFilter
+}
+
+OrderBy input type:
+```graphql
```
Comment (craicoverflow) [inline docs/spec-find.md:29]: You have not closes the last code block.

File: docs/spec-find.md [hunk 25]
```diff
@@ -1,7 +1,84 @@
...
+  in: [String]
+  contains: String
+  startsWith: String
+  endsWith: String
+}
+```
+
+For ID type input will look as follows:
+```graphql
+input IdInput {
```
Comment (craicoverflow) [inline docs/spec-find.md:55]:
```suggestion
input IDInput {
```

File: docs/spec-getOne.md [hunk 26]
```diff
@@ -1,7 +1,14 @@
...
-sidebar_label: spec-getOne
+title: Get Operation
+sidebar_label: Get Operation
 ---
 
-spec-getOne TODO
+## Get single object 
+
+Fetching the note by id can be enabled by specifying id as element. 
+Id can be mapped to id of the object on the resolver level.
```
Comment (craicoverflow) [inline docs/spec-getOne.md:10]: Needs to be reworded. Hard to get.

File: docs/spec-overview.md [hunk 27]
```diff
@@ -4,17 +4,41 @@ title: Overview
...
-Specification 
-
 ## Areas covered
 
 This specification describes all aspects of a flexible GraphQL API suitable for relational databases.
 
-## Focus on API, not implementation/runtime characteristics
+## Focus on API
 
 GraphQLCRUD is a collection of specifications for GraphQL APIs that are designed to work well with specific database technologies. GraphQLCRUD is concerned with the API surface, not the implementation. As such two implementations of GraphQLCRUD could choose to store data in different ways, but applications interacting with the data through the GraphQLCRUD API wouldn't be able to tell the difference.
```
Comment (craicoverflow) [inline docs/spec-overview.md:13]: Should be reworded soon as it from OpenCRUD.

File: docs/spec-overview.md [hunk 28]
```diff
@@ -4,17 +4,41 @@ title: Overview
...
 
 ## SDL
 
 Examples are used throughout this spec to show the final schema generated for a specific data model. In all examples, the SDL notation is used to define the data model. The benefit of SDL is that it is database independent, so we can use the same notation accross all supported databases.
 
+## Naming
+
+GraphQLCRUD does not specify how fields generated for each data type must be named. It is up to each GraphQLCRUD implementation define a naming system. The reference implementation uses the naming convention as listed in the example queries
```
Comment (craicoverflow) [inline docs/spec-overview.md:21]:
```suggestion
GraphQLCRUD does not specify how fields generated for each data type must be named. It is up to each GraphQLCRUD implementation to define a naming system. The reference implementation uses the naming convention as listed in the example queries
```

File: docs/spec-update.md [hunk 29]
```diff
@@ -1,7 +1,55 @@
...
+```graphql
+type Note {
+  id: ID!
+  title: String!
+  description: String
+  comments: [Comment]!
+}
+```
+
+Following mutation will be created:
```
Comment (craicoverflow) [inline docs/spec-update.md:20]: Implies generation.

File: docs/spec-update.md [hunk 30]
```diff
@@ -1,7 +1,55 @@
...
+  # ID field is required for update
+  id: ID!
+  title: String
+  description: String
+}
+```
+
+## Variations
+
+### Contidional updates
```
Comment (craicoverflow) [inline docs/spec-update.md:39]:
```suggestion
### Conditional updates
```

File: docs/spec-update.md [hunk 31]
```diff
@@ -1,7 +1,55 @@
...
+operation only after meeting certain criteria
+
+```graphql
+type Mutation {
+  updateNote(input: MutateNoteInput!, where: UpdateNoteFilter): Note!
+}
+```
+Input type for filtering has all fields marked as optional
+```graphql
+input UpdateNoteFilter {
```
Comment (craicoverflow) [inline docs/spec-update.md:51]: Does this mean batch updates or no?

File: docs/spec-delete.md (hunk 20 above)
Comment (wtrocki) [reply docs/spec-delete.md:20]: Amazing point

File: docs/spec-update.md (hunk 31 above)
Comment (wtrocki) [reply docs/spec-update.md:51]: Actually you are right. Following Hasura docs: 
https://hasura.io/docs/1.0/graphql/manual/schema/basics.html

It is actually single item thing

File: docs/spec-update.md (hunk 10 above)
Comment (wtrocki) [reply docs/spec-update.md:46]: We do not have this implemented but I have put Filter example bellow.

File: docs/schema.graphql (hunk 18 above)
Comment (wtrocki) [reply docs/schema.graphql:1]: Actually I should remove this file :D

File: docs/spec-find.md [hunk 32]
```diff
@@ -1,7 +1,83 @@
...
+```graphql
+input NoteFilter {
+  id: IDInput
+  title: StringInput
+  description: StringInput
+  and: [NoteFilter]
+  or: [NoteFilter]
+  not: NoteFilter
+}
+
```
Comment (wtrocki) [inline docs/spec-find.md:27]:
```suggestion
```
```

File: docs/spec-update.md (hunk 10 above)
Comment (machi1990) [reply docs/spec-update.md:46]: Okay thanks.

File: docs/spec-subscriptions.md (hunk 1 above)
Comment (wtrocki) [reply docs/spec-subscriptions.md:12]: That is subscription filtering. For subscriptions, we do not support advanced filters at the moment. You can filter only by single field. Good point this needws to be explained

File: docs/spec-overview.md (hunk 27 above)
Comment (wtrocki) [inline docs/spec-overview.md:13]:
```suggestion
GraphQLCRUD is a collection of specifications for GraphQL APIs that are abstracting from any database technologies. GraphQLCRUD is concerned with the API only and abstracts from the implementation. As such two implementations of GraphQLCRUD could choose to store data in different ways, but applications interacting with the data through the GraphQLCRUD API wouldn't be able to tell the difference.
```

File: docs/spec-delete.md [hunk 33]
```diff
@@ -1,7 +1,56 @@
...
+```graphql
+type Mutation {
+  deleteNote(input: MutateNoteInput!, where: DeleteNoteFilter): Note!
+}
+```
+Input type for filtering has all fields marked as optional
+```graphql
+input DeleteNoteFilter {
+  title: StringInput
+  description: String
```
Comment (machi1990) [inline docs/spec-delete.md:54]:
```suggestion
  description: StringInput
```

File: docs/spec-subscriptions.md (hunk 1 above)
Comment (machi1990) [reply docs/spec-subscriptions.md:12]: Great.

**Comment by wtrocki** [comment]
Pushed to website. We can refine this over the coming weeks.

**Comment by machi1990** [comment]
Nice work. Thanks :-)

**Pull request merged**